### PR TITLE
Adding first draft of design document

### PR DESCRIPTION
**Pull request by ibeckermayer**
See DESIGN.md

### CODE REVIEW

File: DESIGN.md
Comment (ibeckermayer) [inline DESIGN.md]: Scratch the remark about `Babel`, I was confused re: the need for a [loader](https://webpack.js.org/loaders/) for handling React-style js/jsx and so will be using `Babel` for that.

File: DESIGN.md
Comment (alex-kovoy) [inline DESIGN.md]: >  its default `CookieStore` implementation apparently uses a [non-OWASP-compliant](https://curtisvermeeren.github.io/2018/05/13/Golang-Gorilla-Sessions.html) storage mechanism.

Just wonder what makes it non OWASP compliant ?

File: DESIGN.md
Comment (alex-kovoy) [inline DESIGN.md]: > and set the idle timeout to 5 minutes and the absolute timeout to 1 hour.

Lets ignore the idle timeout. I do not think that it's required. It would be strange to logout a user after any period of inactivity for this app. I would consider implementing timeout mechanism for banking apps but not for this one.

File: DESIGN.md
Comment (alex-kovoy) [inline DESIGN.md]: > Mitigates CSRF attacks

It would prevent only _some_ CSRF attacks, so it should not be considered as a safe protection mechanism alone. For example, if some GET endpoint changes the user state then an attacker can use one-click URL to carry an attack.

File: DESIGN.md
Comment (alex-kovoy) [inline DESIGN.md]: I think you can implement the Use double submit cookies mechanism which would be simpler to implement. Also, could you specify how you are going to pass a CSRF token to the web-client?

File: DESIGN.md
Comment (alex-kovoy) [inline DESIGN.md]: Yes, you can use `rand.Reader.Read(32)` to generate a random string with crypto-strong pseudo random generator. This will give you an opaque token that you can use for session/api keys.

File: DESIGN.md
Comment (alex-kovoy) [inline DESIGN.md:65]: How are you planning to store and verify passwords?

**Review by alex-kovoy — COMMENTED** [review]
I really like your security considerations which you outlined in this document. Good stuff.

One suggestion would be to consider using Access Tokens (for the web) instead of a session cookie. This way you can re-use the same code for API tokens and Access Tokens generation/verification. It also will give you CSRF protection for private endpoints (as you will be passing it in the Authorization header). There will be one public endpoint (login) which is OK because it would be a POST request (so no CSRF protection is required).

You can store AccessTokens in the localStorage on the client side.

This way you do not need to use cookies, csrf tokens, and can reuse a common code which should reduce the scope and still give a reasonable protection for this project.

File: DESIGN.md
Comment (ibeckermayer) [reply DESIGN.md]: The [linked blog post](https://curtisvermeeren.github.io/2018/05/13/Golang-Gorilla-Sessions.html) suggests that session info is stored in the (signed, optionally encrypted) cookie itself, whereas [OWASP](https://cheatsheetseries.owasp.org/cheatsheets/Session_Management_Cheat_Sheet.html) takes for granted that the cookie is a session id that acts as a key to a session management system running on the server.

File: DESIGN.md
Comment (ibeckermayer) [reply DESIGN.md]: Moving away from this design, but the way that the CSRF handoff would have worked is that the client would receive one in an `"X-CSRF-Token"` header when the app first loads, and then would need to return it on subsequent requests to CSRF protected endpoints.

File: DESIGN.md
Comment (ibeckermayer) [reply DESIGN.md:65]: I'm going to use the [bcrypt](https://godoc.org/golang.org/x/crypto/bcrypt) package to save them salted and hashed and then verify against the hash.

File: DESIGN.md
Comment (alex-kovoy) [reply DESIGN.md]: What to store in the cookie is optional, it does not look like that package enforces it. The blog post, it seems, says that if you do not want to store session information in the DB, you can store it in the cookie if your requirements allow it but at the end of the day it's up to you.

Comment (awly) [reply DESIGN.md]: `gorilla/sessions` works with any cookie store that implements https://godoc.org/github.com/gorilla/sessions#Store
The default implementation uses secure cookies, but it's pluggable.

The `scs` library has a nice middleware wrapper, so I see no problem using it.
Just beware of dependency bloat.

**Review by awly — APPROVED** [review]
LGTM on the backend pieces, wait for others to approve the frontend

File: DESIGN.md
Comment (ibeckermayer) [reply DESIGN.md:19]: Ah I see, I was mistaken. @awly [clarifies further](https://github.com/ibeckermayer/teleport-interview/pull/1#discussion_r551586882) below.

File: DESIGN.md
Comment (alex-kovoy) [inline DESIGN.md]: > Using `localStorage` does increase the attack surface for XSS attacks in general

I would not say that it increases the attack surface because XSS targets the rendering layer, un-sanitized inputs, etc. If an attacker managed to run malicious JS in the victim browser then it's already too late.

File: DESIGN.md
Comment (alex-kovoy) [inline DESIGN.md]: I would also add HTML headers to APIs such as:

```
X-XSS-Protection
X-Frame-Options
Strict-Transport-Security
```

**Review by alex-kovoy — APPROVED** [review]
Looks good to me. I would also add HTTP headers to exposed APIs for a better security (or make sure that the library that you are planning to use adds them).

(Please wait for another reviewer to approve as well)

File: DESIGN.md
Comment (ibeckermayer) [reply DESIGN.md]: To clarify, what I was thinking is that it makes the backend security system more vulnerable to successful XSS attacks compared to the previous design with httpOnly cookies, since an attacker couldn't run a script that stole the session-id from an httpOnly cookie but could steal it from localStorage if they knew the right key. Point granted though, without that context my wording/definitions are confusing. Will clarify or remove before merging.

File: DESIGN.md
Comment (ibeckermayer) [reply DESIGN.md]: Good catch. I'll add

```
X-XSS-Protection: 1; mode=block
X-Frame-Options: deny
Strict-Transport-Security: max-age= 6307200; includeSubDomains
Content-Security-Policy: default-src 'self'
X-Permitted-Cross-Domain-Policies: none
```
(Taken from [here](https://wiki.owasp.org/index.php/OWASP_Secure_Headers_Project#tab=Headers))

**Comment by alex-kovoy** [comment]
@ibeckermayer frontend looks good as well

**Pull request merged**
